drivers/rtc/rtc_rv8263.c

```diff
@@ -0,0 +1,918 @@
+    const struct rv8263c8_config *config = p_dev->config;
+    uint8_t value[5];
+
+    if ((p_dev == NULL) || (p_timeptr == NULL) || (id != 0)) {
+        return -EINVAL;
+    }
+
+    (*p_mask) = 0;
+
+    k_spinlock_key_t key = k_spin_lock(&data->lock);
+    err = i2c_burst_read_dt(&config->i2c_bus, RV8263C8_REGISTER_SECONDS_ALARM, value, sizeof(value));
+    k_spin_unlock(&data->lock, key);
+
+    if (err < 0) {
+        LOG_ERR("Error while reading alarm! Error: %i", err);
+        return err;
+    }
+
+    if (value[0] <= MAX_SEC) {
+        p_timeptr->tm_sec = value[0];
+        (*p_mask) |= RTC_ALARM_TIME_MASK_SECOND;
+    }
+
+    if (value[1] <= MAX_MIN) {
+        p_timeptr->tm_min = value[1];
+        (*p_mask) |= RTC_ALARM_TIME_MASK_MINUTE;
+    }
+
+    if (value[2] <= MAX_HOUR) {
+        p_timeptr->tm_hour = value[2];
+        (*p_mask) |= RTC_ALARM_TIME_MASK_HOUR;
+    }
+
+    if (value[3] <= MAX_MDAY) {
+        p_timeptr->tm_mday = value[3];
+        (*p_mask) |= RTC_ALARM_TIME_MASK_MONTHDAY;
+    }
+
+    if (value[4] <= MAX_WDAY) {
+        p_timeptr->tm_hour = value[4];
+        (*p_mask) |= RTC_ALARM_TIME_MASK_WEEKDAY;
+    }
+
+    return 0;
+}
+
+/**
+ * @brief
+ *
+ * @param p_dev         Pointer to device strucure
+ * @param id
+ * @param callback      RTC Alarm callback
+ * @param p_user_data   Pointer to user data for alarm callback
+ * @return int          0 when successful
+ */
+static int rv8263c8_alarm_set_callback(const struct device *p_dev, uint16_t id,
+                                       rtc_alarm_callback callback, void *p_user_data)
+{
+    int err;
+    struct rv8263c8_data *const data = p_dev->data;
+    const struct rv8263c8_config *config = p_dev->config;
+
+    if ((p_dev == NULL) || (id != 0)) {
+        return -EINVAL;
+    }
+
+    k_spinlock_key_t key = k_spin_lock(&data->lock);
+
+    data->alarm_cb = callback;
+    data->alarm_cb_data = p_user_data;
+
+    if (callback != NULL) {
+        err = i2c_reg_update_byte_dt(&config->i2c_bus, RV8263C8_REGISTER_CONTROL_2, RV8263C8_BM_ALARM_INT_ENABLE,
+                                     RV8263C8_BM_ALARM_INT_ENABLE);
+    } else {
+        err = i2c_reg_update_byte_dt(&config->i2c_bus, RV8263C8_REGISTER_CONTROL_2, RV8263C8_BM_ALARM_INT_ENABLE,
+                                     RV8263C8_BM_ALARM_INT_DISABLE);
+    }
+
+    k_spin_unlock(&data->lock, key);
+
+    if (err < 0) {
+        LOG_ERR("Error while writing CONTROL2! Error: %i", err);
+        return err;
+    }
+
+    return 0;
+}
+
+/**
+ * @brief
+ *
+ * @param p_dev     Pointer to device strucure
+ * @param id
+ * @return int      0 when successful
+ */
+static int rv8263c8_alarm_is_pending(const struct device *p_dev, uint16_t id)
+{
+    int ret;
+    struct rv8263c8_data *const data = p_dev->data;
+
+    if ((p_dev == NULL) || (id != 0)) {
+        return -EINVAL;
+    }
+
+    k_spinlock_key_t key = k_spin_lock(&data->lock);
+    ret = data->is_alarm_pending ? 1 : 0;
```

Comment:
Should alarm status testing depend on interrupts? It's not very clear to me from the [API documentation](https://docs.zephyrproject.org/latest/hardware/peripherals/rtc.html#c.rtc_alarm_is_pending), but I would rather do a register read here.

Reply:
I think because this function is part of the alarm API and the alarm API is only available when the alarm is enabled and `int_gpios` is available no test is needed.

Reply:
My understanding is that you can have alarms support enabled without `int_gpios`, so interrupt is only required for the alarm callback.

Reply:
No. Alarms should only work with int-gpios. I thought about enabling it without the GPIOs but it doesn't make sense to me because otherwise the I2C must poll the RTC all the time which occupies the bus.

Reply:
From the `rtc_alarm_set_callback()` documentation:
> When the alarm callback for an alarm is disabled, the alarm triggered event will set the alarm status to “pending”. To check if the alarm status is “pending”, use [rtc_alarm_is_pending()](https://docs.zephyrproject.org/latest/hardware/peripherals/rtc.html#group__rtc__interface_1ga531c379c437a91df44a45e95063700e8).

I ask because my implementation of the RV3028 driver (#73396) may be wrong – I just disable GPIO pin interrupt when there are no callbacks set. Checking the alarm status when calling `rtc_alarm_is_pending()` is only a single I2C read/write. I guess this function isn't for frequent polling and if you need to react quickly to an alarm, you use interrupt with callback. @bjarki-trackunit, what's the correct approach, please?

Reply:
> From the `rtc_alarm_set_callback()` [documentation](https://docs.zephyrproject.org/latest/hardware/peripherals/rtc.html#c.rtc_alarm_set_callback):
> 
> > When the alarm callback for an alarm is disabled, the alarm triggered event will set the alarm status to “pending”. To check if the alarm status is “pending”, use [rtc_alarm_is_pending()](https://docs.zephyrproject.org/latest/hardware/peripherals/rtc.html#group__rtc__interface_1ga531c379c437a91df44a45e95063700e8).
> 
> I ask because my implementation of the RV3028 driver may be wrong – I just disable GPIO pin interrupt when there are no callbacks set.
> 
> https://github.com/zephyrproject-rtos/zephyr/blob/41112f1335e17da5da971d23b343ca0ed4e1188d/drivers/rtc/rtc_rv3028.c#L734-L740
> 
> Checking the alarm status when calling `rtc_alarm_is_pending()` is only a single I2C write/read. I guess this function isn't for frequent polling and if you need to react quickly to an alarm, you use interrupt with callback. @bjarki-trackunit, what's the correct approach, please?

Good question. But why do you disable GPIO interrupt when no callback is set? From my point of view, the alarms should be processed anyway, even if no callback is set. Otherwise, you might have the scenario that an alarm is active because you don´t clear it, then you set a callback and the callback is triggered.

What I mean is, that when you don´t use int-gpios you have to frequently poll the status register to check if an alarm is active and then process it. This will block the bus because a thread (or a worker) is constantly reading the status register. After all, you don´t have an external interrupt. From my perspective using alarms (or updates) without int-gpios isn´t a valid use-case.

But I´m unsure about the `is_pending` API too. Should I clear the source of the alarm in the worker from the interrupt or should the alarm stay active in the RTC until I call `rtc_alarm_is_pending`? I think the current solution isn´t correct because `i2c_reg_update_byte_dt(&config->i2c_bus, RV8263C8_REGISTER_CONTROL_2, RV8263C8_BM_AF, RV8263C8_BM_AF);` must be called from the `rtc_alarm_is_pending` function.

Reply:
The update and alarm callbacks depend on interrupts to work, if `int-gpios` is missing in this case, the user can not set the update or alarm callbacks. In no scenario shall we poll the state of the RTC unless requested as part of an API call like `rtc_get_time()`.

Setting the alarm and testing if it has triggered using `rtc_alarm_is_pending()` shall be possible without `int-gpios`. Once the alarm triggers, the flag is set, and checking the state (and clearing the flag) is done with `rtc_alarm_is_pending()`.

No constant polling :)

The `int-gpios` gpio could be connected to some other external component, triggering a wakeup when the alarm triggersm and checking if the alarm has triggered using `rtc_alarm_is_pending()` can indicate if the alarm was the wakeup source, while also clearing the alarm pending flag :)

Reply:
`rtc_alarm_is_pending()` still relies on the `is_alarm_pending` variable set by the interrupt handler. I would suggest removing this variable and using only the flag in the Control2 register.